transports/noise/src/io/framed.rs

```diff
@@ -43,61 +39,92 @@ static_assertions::const_assert! {
+            session,
             write_buffer: Vec::new(),
             decrypt_buffer: BytesMut::new(),
         }
     }
 
+    fn encode_bytes(&mut self, item: &Vec<u8>, dst: &mut BytesMut) -> Result<(), io::Error> {
+        self.write_buffer
+            .resize(item.len() + EXTRA_ENCRYPT_SPACE, 0);
+        let n = match self.session.write_message(item, &mut self.write_buffer) {
+            Ok(n) => n,
+            Err(e) => {
+                error!("encryption error: {:?}", e);
+                return Err(io::ErrorKind::InvalidData.into());
+            }
+        };
+
+        let prefix = u16::to_be_bytes(n as u16);
+        self.write_buffer.truncate(n);
+
+        dst.put(&prefix[..]);
+        dst.put(&self.write_buffer[..]);
+
+        Ok(())
+    }
+
+    fn decode_bytes(&mut self, src: &mut BytesMut) -> Result<Option<Bytes>, io::Error> {
+        if src.len() < 2 {
+            return Ok(None);
+        }
+
+        let len = u16::from_be_bytes([src[0], src[1]]) as usize;
+        if len == 0 || src.len() < len + 2 {
+            return Ok(None);
+        }
+
+        src.advance(2);
+        self.decrypt_buffer.resize(len, 0u8);
+        let n = match self
+            .session
+            .read_message(&src[0..len], &mut self.decrypt_buffer)
+        {
+            Ok(n) => n,
+            Err(e) => {
+                debug!("decryption error {e}");
+                return Err(io::ErrorKind::InvalidData.into());
+            }
+        };
+
+        self.decrypt_buffer.truncate(n);
+        src.advance(len);
+
+        Ok(Some(self.decrypt_buffer.split().freeze()))
+    }
+}
+
+impl Codec<snow::HandshakeState> {
+    /// Checks if the session was started in the `initiator` role.
     pub(crate) fn is_initiator(&self) -> bool {
         self.session.is_initiator()
     }
 
+    /// Checks if the session was started in the `responder` role.
     pub(crate) fn is_responder(&self) -> bool {
         !self.session.is_initiator()
     }
 
-    /// Converts the `NoiseFramed` into a `NoiseOutput` encrypted data stream
-    /// once the handshake is complete, including the static DH [`PublicKey`]
-    /// of the remote, if received.
+    /// Converts the underlying Noise session from the `Handshake` state to a
+    /// `Transport` state once the handshake is complete, including the static
+    /// DH [`PublicKey`] of the remote if received.
     ///
-    /// If the underlying Noise protocol session state does not permit
-    /// transitioning to transport mode because the handshake is incomplete,
-    /// an error is returned. Similarly if the remote's static DH key, if
-    /// present, cannot be parsed.
-    pub(crate) fn into_transport(self) -> Result<(PublicKey, Output<T>), Error> {
+    /// If the Noise protocol session state does not permit transitioning to
+    /// transport mode because the handshake is incomplete, an error is returned.
+    /// Similarly if the remote's static DH key, if present, cannot be parsed.
```

Comment:
We could expand here that we _only_ support the XX protocol in libp2p and thus, not having a static DH key is a fatal handshake error: https://github.com/libp2p/specs/tree/master/noise#design-considerations